### PR TITLE
Fixing RECO in pp_on_AA era jet reconstruction

### PR DESCRIPTION
**Pull request by stepobr**
This pull request is a follow-up of https://github.com/cms-sw/cmssw/pull/24380
The issue is partly described in https://github.com/cms-sw/cmssw/issues/24587
The ideal scenario for https://github.com/cms-sw/cmssw/pull/24380 would be to get rid of all jet algorithms not used for pp_on_AA era. However 3 of them (ak4PF, ak4PFCHS, ak8PFCHS) were kept because otherwise it failed to pass the validation step. But in that case it was later found that the sequences crashes if stopped after RECO step due to the absence of pfNoPileUpJME used by the 3 mentioned algorithms. Trivial solution would be to add back pfNoPileUpJME. But in order to speed-up reconstruction it was replaced by a dummy EDFilter with a very high pt threshold which results in an empty input for these 3 jet modules. The filter does not prevent the modules to be run, but since they have no input objects they do nothing.

@cfmcginn @mandrenguyen

### CODE REVIEW

File: RecoJets/JetProducers/python/ak4PFJets_cfi.py [hunk 1]
```diff
@@ -33,3 +33,6 @@
     doAreaFastjet = cms.bool(True),
     jetPtMin = cms.double(100.0)
     )
+from Configuration.Eras.Modifier_pp_on_AA_2018_cff import pp_on_AA_2018
+pp_on_AA_2018.toModify(ak4PFJets,src = cms.InputTag("pfNoPileUpJMEHI"), inputEtMin = cms.double(9999))
+pp_on_AA_2018.toModify(ak4PFJetsCHS,src = cms.InputTag("pfNoPileUpJMEHI"), inputEtMin = cms.double(9999))
```
Comment (slava77) [inline RecoJets/JetProducers/python/ak4PFJets_cfi.py]: please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
Just to better understand the logic (and have it reported in this  github thread for the record):
- To speed up the jet reconstruction in the HI sequence you want to remove 
```
                           ak4PFJets,
                           ak4PFJetsCHS,
                           ak8PFJetsCHS,
                           PFTowers,			   
                           akPu3PFJets,
                           akPu4PFJets,
                           kt4PFJetsForRho,
                           hiFJRhoProducer,
                           akCs4PFJets
```
from the sequence itself.
- You can't simply do so because as such you would mess up the DQM
- Instead of removing the unwanted module from the reco sequence and adjusting the DQM for HI accordingly, you prefer to add a "filter" which selects an empty set of pf objects to be sent to the jet producers.
- "Real" filters are not allowed in reco sequences (in fact the original PFNoPilupJME was an EDProducer, not an EDFilter). Your new PFNoPilupJMEHI is an EDFIlter, but it does not filter, in practice; it only selects objetcts. As such it can enter in the reco sequence: it does not prevent the subsequent modules to be run, but since they get no input objects after the selection they do nothing.

Please confirm that my understanding is correct, or correct it if not.
Please also write the whole logic behind this "fix" in the PR description, so that it remains available here.

**Comment by stepobr** [comment]
Most of your points are correct, a few comments:

- May be that's a typo but from that list ideally I would get rid of only ak4PFJets, ak4PFJetsCHS, ak8PFJetsCHS.
- I can't do that not because of DQM, but rather because of the validation step. For example, excluding ak4PFJets from the sequence leads to:
`An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 3
   [1] Running path 'prevalidation_step'
   [2] Prefetching for module MultiTrackValidator/'trackValidator'
   [3] Prefetching for module JetTracksAssociationToTrackRefs/'cutsRecoTracksAK4PFJets'
   [4] Calling method for module JetTracksAssociatorExplicit/'ak4JetTracksAssociatorExplicitAll'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for a container with elements of type: reco::Jet
Looking for module label: ak4PFJets
Looking for productInstanceName:
`
- Yes, this is the correct logic behind using this filter. 

I will update the description now.

**Comment by perrotta** [comment]
@cmsbuild please test
(to have back the plots with differences, which have disappeared)

**Comment by perrotta** [comment]
+1
- ak4PF, ak4PFCHS, ak8PFCHS are "sterilized" thanks to the pfNoPileUpJME selector, which feeds them with an empty set of PF objects
- The implementation follows what is in the PR description
- Changes in the jenkins tets outputs are only in the 2018 HI workflow, and they correspond to the expectations (empty ak4PF, ak4PFCHS, ak8PFCHS collections)

**Comment by fabiocos** [comment]
+1

**Pull request merged**